org.eclipse.xtext/src/org/eclipse/xtext/nodemodel/impl/GrammarElementsInterner.java

```diff
@@ -0,0 +1,100 @@
+ * The node model contains nodes derived from {@link AbstractNode}.  An {@link AbstractNode} has an array describing its grammar 
+ * elements.  This array is often identical for many {@link AbstractNode}s.  There can be millions of nodes in a node model
+ * so it makes sense to remove the duplicate arrays.  This {@link GrammarElementsInterner} tries to reuse an interned version of such an array
+ * wherever possible
+ */
+
+public class GrammarElementsInterner {
+
+	private static class InternKey {
+		final EObject grammarElement;
+		final Object grammarElements;
+		
+		int hashCode = -1;
+
+		InternKey(EObject grammarElement, Object grammarElements) {
+			this.grammarElement = grammarElement;
+			this.grammarElements = grammarElements;
+		}
+		
+		EObject[] prefixElement() {
+			if(grammarElements instanceof EObject) {
+				return new EObject[] {grammarElement, (EObject) grammarElements};
+			} else {
+				return ObjectArrays.concat(grammarElement, (EObject[]) grammarElements);
+			}
+		}
+		
+		EObject[] postfixElement() {
+			if(grammarElements instanceof EObject) {
+				return new EObject[] {(EObject) grammarElements, grammarElement };
+			} else {
+				return ObjectArrays.concat((EObject[]) grammarElements, grammarElement);
+			}
+		}
+
+
+		@Override
+		public int hashCode() {
+			if (hashCode == -1)
+				hashCode = Objects.hash(grammarElement, grammarElements);
+			return hashCode;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if(o instanceof GrammarElementsInterner.InternKey) {
+				GrammarElementsInterner.InternKey interned = (GrammarElementsInterner.InternKey) o;
+				return Objects.equals(this.grammarElement, interned.grammarElement) 
+						&& Objects.equals(this.grammarElements, interned.grammarElements);
+			}
+			
+			return false;
+		}
+	}
+
+	Map<GrammarElementsInterner.InternKey, EObject[]> interningMap = Maps.newHashMap();
+	
+	/**
+	 * Take a list of grammar elements and add an additional grammar element to the front.  If this
+	 * list has already been produced in the past, return an interned version.  Note that 
+	 * this code will only function efficiently if repeated invocation uses the same grammar element
+	 * objects and arrays or else a new concatenated array will be created every time.
+	 */
+	EObject[] prependAndIntern(EObject grammarElement, Object grammarElements) {
+		GrammarElementsInterner.InternKey internKey = new InternKey(grammarElement, grammarElements);
+		return interningMap.computeIfAbsent(internKey, key -> key.prefixElement());
+	}
+	
+	/**
+	 * Take a list of grammar elements and add an additional grammar element at the back.  If this
+	 * list has already been produced in the past, return an interned version.  Note that 
+	 * this code will only function efficiently if repeated invocation uses the same grammar element
+	 * objects and arrays or else a new concatenated array will be created every time.
+	 */
+	EObject[] appendAndIntern(Object grammarElements, EObject grammarElement) {
```

Comment:
This could be strongly typed

Reply:
I tried a bit but the resulting code is quite asymetric.  The problem is that in some case I indeed know that the grammarElements are an EObject [] but in some other cases not.  I prefer keeping it as is.

Reply:
Since it's package visible and in nodemodel.impl I'm ok with the signatures, too.